### ISSUE TITLE
[Feature request]: Undo/Redo Support

### ISSUE THREAD

**Issue by arjo129** [open-rmf/rmf_site#245]
### Before proceeding, is there an existing issue or discussion for this?

- [X] I have done a search for similar [issues](.?q=type%3Aissue) and [discussions](https://github.com/open-rmf/rmf/discussions).

### Description

As mentioned in the title. This will likely need a major refactor :skull_and_crossbones: , in traffic editor we were fortunate to have Qt's built in `QAction` based undo/redo workflow. For Site Editor, we should support some way of undoing a change.

### Implementation Considerations

There are several ways to support undo/redo. The naivest way is to snapshot the current world state and keep it in a buffer. This will not scale well as worlds get more complex.

A less naive way would be to require that every forward edit has an equivalent inverse edit. For instance in Qt, this is exactly what the QAction class does. TBF, I'm not sure of the best way we would implement this particularly, with the context of `bevy_impulse` in mind.

Special care will have to be taken when we handle moving and deletion of objects as well.

### Alternatives

Saving intermediate `.ron` files :skull_and_crossbones: .

**Comment by luca-della-vedova** [comment]
If I remember correctly, one of the intentions behind using the [ChangePlugin](https://github.com/open-rmf/rmf_site/blob/main/rmf_site_editor/src/site/change_plugin.rs) for submitting changes to components rather than writing directly to their value was exactly for this reason, to be able to intercept them and save an undo redo buffer (there is a [comment](https://github.com/open-rmf/rmf_site/blob/main/rmf_site_editor/src/site/change_plugin.rs#L23-L25) about it).

Note however that the `Change` plugin does _not_ cover everything. Specifically entity creation and deletion does not go through a `Change` plugin. However I can imagine intercepting the `Delete` events to undo / redo deletion and keeping track of what was created to undo creation.
There might also be other component changes that don't go through a `Change` plugin but I'm not 100% sure

**Comment by arjo129** [comment]
Yeah I do believe that was one of the reasons to use the change plugin, IIRC that it does not cover movement (which is often the thing I want to undo most)
